### DISCUSSION TITLE
How to get the underlying `table` instance?

### DISCUSSION THREAD

**Discussion by Janekk** [KevinVandy/material-react-table#23] [Q&A] [answered]
I'd like to get a `table` instance (`MRT_TableInstance`) in order to use its api method `getSelectedRowModel`, what is the best way to do this?

**Comment by KevinVandy** [comment]
So if you are already following [this example](https://www.material-react-table.com/docs/guides/row-selection#enable-row-selection-source-code) and managing the `rowSelection` state yourself, you should already have access to all the selected rows.

All `render[Section]` props should already have access to the tableInstance if passed a function.

For example, `renderToolbarTopActions={({ table }) => <div>...</div>}`

Hmm I guess we could add a prop like `onTableInstanceUpdated` or something like that. Though it would kind of be readonly.

Is the use case for some buttons/components outside the table?

**Reply by Janekk** [reply]
Thank you.
It seems like `rowSelection` state only has the id of a selected row, I'd like to access the underlying row's data though. I was thinking about using `getSelectedRowModel` for that.

`onTableInstanceUpdated` for read-only purposes could work :)

**Reply by KevinVandy** [reply]
Alright I think I'll add it, but will need to investigate to make sure it does not have any performance issues that it creates.

Added as of v0.26.3 - `onTableInstanceChange` prop

**Comment by Janekk** [comment]
btw. for version >= 0.26 I'm getting `TypeError: Cannot read properties of undefined (reading 'charAt')` from MRT_TableHeadCellFilterLabel
![image](https://user-images.githubusercontent.com/6827881/179610571-72d6aca8-2391-4ec0-a1c9-325ea98c9ebd.jpg)

**Reply by KevinVandy** [reply]
Which filterFn are you using?

**Reply by Janekk** [reply]
actually the issue starts with https://github.com/KevinVandy/material-react-table/commit/1eabb50a4b2df39517cad018722ff0b6b81ccad9#diff-6a8aef832f51ea5fdf1555f9570eb2729e776511099e5f4bc5b15f75059fba87 I'll try to isolate the error today

**Reply by KevinVandy** [reply]
I added more optional chaining there in v0.26.4 so it at least should not crash. More filtering and custom filtering improvements to come

**Reply by Janekk** [reply]
@KevinVandy it fixed the issue, thank you! :tada:

**Comment by KevinVandy** [comment]
@Janekk Hi, if you see this, I'm considering removing the onTableInstanceChange prop from Material React Table, as it can easily cause infinite renders when used the wrong way. I wanted to ask for your use case to make sure there was a working alternative for you.

**Reply by Janekk** [reply]
Hi @KevinVandy, no problem I don't actually have a use case for it now. Thank you

**Comment by KevinVandy** [comment]
There is now actually a new stable way to get the underlying table instance with the `tableInstanceRef` prop. More info to come soon, but it will not cause infinite renders, or any re-renders for that matter, but it will allow you to have access to the tableInstance in any scope of your codebase. Introduced in `v0.40.0`

**Comment by KevinVandy** [comment, marked as answer]
Now in v2, this is trivial with the `useMaterialReactTableProps` https://www.material-react-table.com/migrating-to-v2#usematerialreacttable
